File: akka_essentials/src/main/java/essentials/actor/java/chapter2/actor/MasterActor.java
package essentials.actor.java.chapter2.actor;

import essentials.actor.java.chapter2.message.MapData;
import essentials.actor.java.chapter2.message.ReduceData;
import essentials.actor.java.chapter2.message.Result;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.routing.RoundRobinPool;

public class MasterActor extends UntypedActor {

    private ActorRef mapActor = context().actorOf(Props.create(MapActor.class, MapActor::new).withRouter(new RoundRobinPool(5)), "map");
    private ActorRef reduceActor = context().actorOf(Props.create(ReduceActor.class, ReduceActor::new)
            .withRouter(new RoundRobinPool(5)), "reduce");
    private ActorRef aggregateActor = getContext().actorOf(Props.create(AggregateActor.class, AggregateActor::new), "aggregate");

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof String) {
            mapActor.tell(message, self());
        } else if (message instanceof MapData) {
            reduceActor.tell(message, self());
        } else if (message instanceof ReduceData) {
            aggregateActor.tell(message, self());
        } else if (message instanceof Result) {
            aggregateActor.forward(message, getContext());
        } else
            unhandled(message);
    }

}

File: akka_essentials/src/main/scala/essentials/actor/scala/chapter2/actor/MasterActor.scala
package essentials.actor.scala.chapter2.actor


import akka.actor.{Actor, Props}
import akka.routing.RoundRobinPool
import essentials.actor.scala.chapter2.{MapData, ReduceData, Result}


class MasterActor extends Actor {
  val mapActor = context.actorOf(Props[MapActor].withRouter(RoundRobinPool(nrOfInstances = 5)), name = "map")
  val reduceActor = context.actorOf(Props[ReduceActor].withRouter(RoundRobinPool(nrOfInstances = 5)), name = "reduce")
  val aggregateActor = context.actorOf(Props[AggregateActor], name = "aggregate")

  def receive: Receive = {
    case line: String => mapActor ! line
    case mapData: MapData => reduceActor ! mapData
    case reduceData: ReduceData => aggregateActor ! reduceData
    case Result => aggregateActor forward Result
  }

}
